src/Juvix/Compiler/Core/Transformation/MatchToCase.hs

```diff
@@ -1,338 +1,314 @@
-combineCompiledPatterns ps = go indexedPatterns
-  where
-    indexedPatterns :: [Indexed (Sem ((Reader CompileStateNode) ': r) CompiledPattern)]
-    indexedPatterns = reverse (indexFrom 0 (reverse ps))
-
-    go :: [Indexed (Sem ((Reader CompileStateNode) ': r) CompiledPattern)] -> Sem r CompiledPattern
-    go [] = asks (^. compileStateCompiledPattern)
-    go (Indexed depth cp : xs) = do
-      numBinders <- length <$> asks (^. compileStateCompiledPattern . compiledPatBinders)
-      nextPattern <- runReader (CompileStateNode (mkVar' (numBinders + depth))) cp
-      updateState nextPattern (go xs)
+        go :: Int -> [(Node, Type)] -> Sem r Node
+        go n = \case
+          [] ->
+            compile err n [0 .. n - 1] matrix
+            where
+              err = hsep . take n
+              matrix = map matchBranchToPatternRow _matchBranches
+
+              matchBranchToPatternRow :: MatchBranch -> PatternRow
+              matchBranchToPatternRow MatchBranch {..} =
+                PatternRow
+                  { _patternRowPatterns = toList _matchBranchPatterns,
+                    _patternRowBody = _matchBranchBody,
+                    _patternRowIgnoredPatternsNum = 0,
+                    _patternRowBinderChangesRev = [BCAdd n]
+                  }
+          (val, valty) : vs' -> do
+            ty' <- goMatchToCase (recur . (BCAdd n :)) valty
+            val' <- goMatchToCase (recur . (BCAdd n :)) val
+            mkLet' ty' val' <$> go (n + 1) vs'
+
+    -- `compile err bindersNum vs matrix`:
+    --  - `err` creates a textual representation of an unmatched pattern
+    --    sequence, given as arguments the representations of patterns for the
+    --    holes (corresponding to the matched values `vs`)
+    --  - `bindersNum` is the number of binders added so far
+    --  - `vs` are the de Bruijn levels of the values matched on w.r.t. the
+    --    first added binder; a value matched on is always a variable referring
+    --    to one of the binders added so far
+    --  - `matrix` is the pattern matching matrix
+    compile :: ([Doc Ann] -> Doc Ann) -> Level -> [Level] -> PatternMatrix -> Sem r Node
+    compile err bindersNum vs matrix = case matrix of
+      [] -> do
+        -- The matrix has no rows -- matching fails (Section 4, case 1).
+        fCoverage <- asks (^. optCheckCoverage)
+        if
+            | fCoverage ->
+                throw
+                  CoreError
+                    { _coreErrorMsg = "Pattern matching not exhaustive. Example pattern sequence not matched: " <> show (ppOutput $ err (repeat "_")),
+                      _coreErrorNode = Nothing,
+                      _coreErrorLoc = fromMaybe defaultLoc (getNodeLocation node)
+                    }
+            | otherwise ->
+                return $
+                  mkBuiltinApp' OpFail [mkConstant' (ConstString ("Pattern sequence not matched: " <> pat))]
+        where
+          pat = show (ppOutput $ err (repeat "_"))
```

Comment:
could we have `replicate ? "_"` instead of `repeat`?

Reply:
Done